### PR TITLE
Make pkgs.nixos and pkgs.nixosTest pure wrt NIXOS_EXTRA_MODULE_PATH

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

Includes #51850

This avoids the `NIXOS_EXTRA_MODULE_PATH` impurity in NixOS when
Nixpkgs is invoked 'purely' via `pkgs/top-level/default.nix`, but
passed the impurity along if invoked impurely via `/default.nix`
or `pkgs/top-level/impure.nix`.

`pkgs.nixosTests` remains impure by necessity until a possible future
optimization that will make it call `pkgs.nixosTest`, therefore
inheriting its behavior.

This duplicates the default in a number of places because these files are potentially public interfaces.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date (NIXOS_EXTRA_MODULE_PATH has minimal documentation in NixOS manual which remains correct)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/lib/build-vms.nix [hunk 1]
```diff
@@ -6,7 +6,8 @@
   # Nixpkgs, for qemu, lib and more
 , pkgs
 , # NixOS configuration to add to the VMs
-  extraConfigurations ? []
+  extraConfigurations ? let e = builtins.getEnv "NIXOS_EXTRA_MODULE_PATH";
+                        in if e == "" then [] else [(import e)]
```
Comment (grahamc) [inline nixos/lib/build-vms.nix]: Hrm, a bit ugly to proliferate this code around. Are there nicer ways to accomplish this?

File: pkgs/top-level/impure.nix [hunk 2]
```diff
@@ -72,15 +72,24 @@ in
         else overlays homeOverlaysDir
       else []
 
+  # An impurity in NixOS to pass on when invoking pkgs.nixos,
+  # pkgs.nixosTest, pkgs.nixosTests.
+, nixosExtraModules ? let e = builtins.getEnv "NIXOS_EXTRA_MODULE_PATH";
```
Comment (Ericson2314) [inline pkgs/top-level/impure.nix]: I would maybe call this `nixosTestExtraModules` so it doesn't just sound like some layer violation.

File: nixos/lib/build-vms.nix (hunk 1 above)
Comment (Ericson2314) [reply nixos/lib/build-vms.nix]: Maybe we can just not put it most places, and then see what breaks :D

**Comment by ottidmes** [comment]
I completely agree with the reason for these changes, namely that you should be able to make your builds completely pure and that the choice should not be between impure and less impure, rather it should be between pure and impure. I do have two points of feedback though:

1. Ideally we should have pure by default and only add impurities explicitly, while these changes make it from always potentially impure to potentially impure by default, so an improvement, but still not what we should aim for. That does not mean that I am against merging these changes, because it is still an improvement in the meantime.

2. From checking your changes, it seems like you got it covered, but just in case I will ask: is the original use case for adding this environment variable still covered? [The ability to change _every_ NixOS evaluation externally with the use of an environment variable.](https://github.com/NixOS/nixpkgs/commit/73f18fd42f6d266f586c4b80b596a0f761a904df#diff-e29cca50467118cb72b2d2c8f97cd7ca) With the exception of the NixOS packages not run through `impure.nix`, but that is a good thing and does not break the use case.

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Comment by roberth** [comment]
This PR was too conservative. Nowaday's, we're moving toward a Nix that doesn't read environment variables by default, so we'll probably remove this feature completely at some point in the future.

**Pull request closed without merging**